File: src/js/core/src/Style/ComponentVariant.ts
import type { StyleDefinitionComponentVariant } from '../StyleDefinition.js';
import { Element } from './Element.js';

/**
 * Read-only view over an entry in a component's `variants` block.
 */
export class ComponentVariant {
  #data: StyleDefinitionComponentVariant;
  #elements?: readonly Element[];

  constructor(data: StyleDefinitionComponentVariant) {
    this.#data = data;
  }

  /**
   * Returns the variant's elements, lazily wrapped as {@link Element}
   * instances on first access.
   */
  elements(): readonly Element[] {
    this.#elements ??= this.#data.elements.map((el) => new Element(el));

    return this.#elements;
  }

  /**
   * Returns the weighted-pick weight for this variant, defaulting to `1`.
   */
  weight(): number {
    return this.#data.weight ?? 1;
  }
}

File: src/dart/core/lib/src/style/component_variant.dart
/// Read-only view over an entry in a component's `variants` block.
library;

import 'element.dart';

/// A single variant of a component: the element subtree it renders and its
/// weighted-pick weight.
class ComponentVariant {
  final Map<String, Object?> _data;
  List<ElementNode>? _elements;

  ComponentVariant(Map<String, Object?> data) : _data = data;

  /// Returns the variant's elements, lazily wrapped as [ElementNode]
  /// instances on first access.
  List<ElementNode> elements() => _elements ??= List.unmodifiable(
        (_data['elements'] as List<Object?>)
            .map((element) => ElementNode(element as Map<String, Object?>)),
      );

  /// Returns the weighted-pick weight for this variant, defaulting to `1`.
  double weight() {
    final weight = _data['weight'];

    return weight == null ? 1 : (weight as num).toDouble();
  }
}
